### PR TITLE
SSHClient.connect: support two factor callback

### PR DESCRIPTION
**Pull request by g-braeunlich**
This commit adds a new argument `two_factor_callback` to SSHClient.connect.
This way, it is possible to call this function in a gui, when you do not know in advance (cannot use the password argument), if there a 2nd factor is required and you want to get the users input from something different than the terminal.

### CODE REVIEW

File: paramiko/client.py [hunk 1]
```diff
@@ -383,7 +388,7 @@ def connect(
         )
         if self._log_channel is not None:
             t.set_log_channel(self._log_channel)
-        if banner_timeout is not None:
+            if banner_timeout is not None:
```
Comment (ploxiln) [inline paramiko/client.py]: bad indent here

Comment (g-braeunlich) [reply paramiko/client.py]: thx for the careful review!
Should be fixed now

File: paramiko/client.py [hunk 2]
```diff
@@ -314,6 +315,10 @@ def connect(
         :param dict disabled_algorithms:
             an optional dict passed directly to `.Transport` and its keyword
             argument of the same name.
+        :param two_factor_callback:
+            an optional callable with no arguments. If specified, it will be
+            called when a 2nd factor is needed. Usefull when calling connect
+            from a gui which needs to open a dialog
```
Comment (allanlewis) [inline paramiko/client.py]: A few minor things here:

- `s/Usefull/Useful/`
- Please surround "connect" with backticks as it's a code object.
- `s/gui/GUI/`
- Please append a full stop/period.

Comment (g-braeunlich) [reply paramiko/client.py]: Thx. Done

**Comment by g-braeunlich** [comment]
Anything else to improve here?

**Comment by g-braeunlich** [comment]
anything I can do to get this merged?

**Pull request closed without merging**